lib/spack/spack/build_environment.py

```diff
@@ -128,8 +133,16 @@ def __call__(self, *args, **kwargs):
         parallel = not disable and kwargs.get('parallel', self.jobs > 1)
 
         if parallel:
-            jobs = "-j%d" % self.jobs
-            args = (jobs,) + args
+            jobs = kwargs.pop('jobs_arg', '-j{0}')
```

Comment:
IMO the following would be more straightforward (I don't see anything setting the new `jobs_arg` variable so IMO it isn't needed):

```
jobs_arg = 'j{0}'.format(self.jobs)
args = (jobs_arg,) + args
if jobs_env:
  ...
```

(I also think this avoids setting multiple jobs arguments, which may be valid and appears to be handled here, but I'm not aware of any system that needs it so I'd prefer to avoid the complexity)

Does it make sense to avoid adding a `-jN` argument when `jobs_env` is set?

I think it's worth adding a note that you are using `kwargs.pop` because `jobs_env` is intended for use by this function specifically and that the remaining `kwargs` are to be passed to the `make` invocation.

Reply:
`jobs_arg` is there for a little more flexibility for new `make`-like execs, for wrappers using long-form options (`ctest` has a `--parallel` alias, for example) or separating the `N` from the `-j`, and for the case where one actually does want to avoid `-jN` when `job_env` is set. Note that in the case `jobs_env` was put in place to address (`make test` for CMake projects) we can't assume for the general CMake case because any given project can provide its own `make test` target which may or may not honor `-jN`. @davydden mentioned `dealii` as a package for which this is done.

I can certainly add a note about the use of `kwargs.pop()`. I'll defer to you on whether I should go for simplicity or flexibility on `jobs_arg` -- let me know what you think after my explanation.

Reply:
Is there any known case where we can't just supply a flag? Like could we do `make -j 4` instead of `make -j4`? That would make things a bit simpler in my mind. I suppose there could be a flag that means "use all cores" and no option to supply a specific number. But I think that would break this anyway?

Reply:
@adamjstewart The generated test target for CMake projects requires the `CTEST_PARALLEL_LEVEL` environment variable to be set for parallelism (ignoring the `-j` option to `make`), which is the whole reason for this PR. Different `make`-like applications handle non-specific parallelism differently anyway: `ninja` defaults to core-level parallelism without any -j, whereas `make` omits the number for "infinite" parallelism. It's messy. I could always go back to the old system of assuming `-j{0}` in the absence of use cases, at the cost of flexibility for the future. Is that the consensus?

Reply:
Personally, I would be fine with the previous behavior, the proposed behavior, or the behavior I'm suggesting. To be clear, what I'm suggesting is replacing `jobs_arg='-j{0}'` with `jobs_flag='-j'`. The only difference from our current behavior would be adding a space after the `-j`.

But if you want to avoid debate and just get this merged, I would avoid making changes to the UI that aren't absolutely necessary until we find a use case for those changes. Although @alalazo and I are generally fans of making things as generic as possible...

Reply:
> for wrappers using long-form options (ctest has a --parallel alias, for example)

Like a wrapper around `ctest`, called `ctest`, which only accepts `--parallel`?

> separating the N from the -j

Do any systems require this?

> and for the case where one actually does want to avoid -jN when job_env is set

How does one accomplish that? e.g. if `jobs_arg=None` the next line fails. The only way I see to do that now is to pass an empty iterable, which IMO is not straightforward.

Also, I didn't see any of those cases explaining the handling of `jobs_arg` as a tuple (other than my inference on the last one)

Overall I'd be fine if `jobs_arg` was a single argument like `jobs_env`.

Reply:
I have removed the currently-unneeded flexibility, leaving us with the minimum changes required to provide parallel test execution under CMake.